rest_framework_json_api/views.py

```diff
@@ -62,10 +70,52 @@ class MyViewSet(viewsets.ModelViewSet):
         return qs
 
 
+class SelectAndPrefetchForIncludesMixin(object):
+    """
+    This mixin provides a helper attributes to select or prefetch related models
+    based on the include specified in the URL.
+
+    __all__ can be used to specify a prefetch which should be done regardless of the include
+
+    .. code:: python
+
+    # When MyViewSet is called with ?include=author it will prefetch author and authorbio
+    class MyViewSet(viewsets.ModelViewSet):
+        queryset = Book.objects.all()
+        prefetch_for_includes = {
+            '__all__': [],
+            'category.section': ['category']
+        }
+        select_for_includes = {
+            '__all__': [],
+            'author': ['author', 'author__authorbio'],
+        }
+    """
+    def get_queryset(self):
+        qs = super(SelectAndPrefetchForIncludesMixin, self).get_queryset()
+
+        includes = self.request.GET.get('include', '').split(',') + ['__all__']
+
+        if hasattr(self, 'select_for_includes'):
+            selects = [self.select_for_includes.get(inc) for inc in includes]
+            qs = qs.select_related(*selects)
+
+        if hasattr(self, 'prefetch_for_includes'):
+            prefetches = [self.prefetch_for_includes.get(inc) for inc in includes]
+            qs = qs.prefetch_related(*prefetches)
+
+        return qs
+
+
 class AutoPrefetchMixin(object):
     def get_queryset(self, *args, **kwargs):
         """ This mixin adds automatic prefetching for OneToOne and ManyToMany fields. """
         qs = super(AutoPrefetchMixin, self).get_queryset(*args, **kwargs)
+
+        # Prefetch includes handled by another mixin, let's do not mix them
+        if hasattr(self, 'prefetch_for_includes'):
```

Comment:
This is not optimal as if I simply want to add one prefetch to one include but leave the rest I cannot do it. It is also not backwards compatible.

If I want to I do not need to derive from both mixins so I would simply leave this out.

Reply:
Agree, removed the change